A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+    *   The sum of
+        [weights](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L278)
+        must add up to the
+        [total\_weight](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L335).
+*   Can be
+    [cluster\_header](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L712)
+    *   Note: the route containing this action will be ignored. This feature can
+        be accomplished by header matching.
+
+Duplicate matches are allowed, but because the first match wins, the second
+will never be used.
+
+Violating any of the validation causes the response to be rejected (NACK’ed).
+The fields not covered by this section will all be ignored.
+
+### Service config generating
+
+The service config generated by the xDS resolver will pick
+`xds_routing_experimental` as the LB policy. As an optimization, when there’s
+only one route, the LB policy picked will be `cds_experimental` or
+`weighted_target_experimental` depending on the action, so we can skip the
+`xds_routing` policy that has only one child.
+
+The `xds_routing_experimental` balancer config will contain the validated
+`{matchers, action}` pairs. The actions will be in a separate list, and will
+be referred to by the routes by their names.
+
+#### Action list
+
+All the actions from the validated `{matchers, action}` pairs will be
+de-duplicated and stored in a cache. All actions in the cache will be
+assigned a unique name.
+
+When processing an RDS response, for each action:
+
+*   If the action is cluster, the child\_policy will be `cds_experimental`, with
+    a balancer config containing the cluster string.
+*   If the action is weighted\_clusters, the child\_policy will be
+    `weighted_target_experimental`. The balancer config will contain a list of
+    `cds_experimental` LB configs, and their corresponding weights.
+*   If this action is already in the cache, it will be reused, and nothing needs
+    to be done
+*   If the action is new, it will be be added to the list and assigned a name
+    *   The name should be a human readable hash of the config, to be logged for
+        debugging purposes.
+    *   For example, the names can be `cds:cluster1` and
+        `weighted:cluster1_cluster2_<index_number>`.
+
+When assigning names to new actions, we will try to reuse the existing names.
+This will help the xds\_routing policy to not always close/create child
+policy. But we will limit the name reuse only between weighted cluster
+actions that share the same set of clusters (the only update is weights).
+This covers the typical use case where the user keeps updating the cluster
+weights during a canary rollout.
+
+To do this, for each new RDS response, the client will
+
+1.  Calculate the list of new actions to add, and the list of old actions to remove.
```

Comment:
I think this can wind up being a simple two-pass algorithm that generates a new map based on the new list of WeightedClusters actions and the old map:

1. Pass 1: For each new WeightedClusters action, if there is an exactly matching action in the old map, move it from the old map to the new map.
2. Pass 2: For each new WeightedClusters action to add that did not match in pass 1, if there is an action in the old map that has the exact same list of clusters but different weights, move it from the old map to the new map.  Otherwise, add a new action name to the new map.

Note that the above needs to be done for WeightedClusters actions only.  Actions that directly specify the cluster do not need to go through this algorithm at all.

This is what we implemented in C-core: https://github.com/grpc/grpc/blob/832da8bd26ce4617adf667837f30d664ed191187/src/core/ext/filters/client_channel/xds/xds_client.cc#L2141

Reply:
I think is in sync with what the design is proposing. Pass 1 produces 3 lists: `new-actions-to-add`, `actions-that-didn't-change`, and `actions-to-remove`. Pass 2 tries to reuse `actions-to-remove` for `new-actions-to-add`.

I removed the pseudocode because it has unnecessary details, and is just causing confusion.